modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java

```diff
@@ -475,6 +555,85 @@ private void update(Series series) throws SearchIndexException {
     }
   }
 
+  /**
+   * Adds or updates the series in the search index. Uses a locking mechanism to avoid issues like Lost Update.
+   *
+   * @param ids
+   *          The ids of the series to add
+   * @param updateFunctionBundle
+   *          The functions that do the actual updating
+   * @param orgId
+   *          The organization the series belongs to
+   * @param user
+   *          The user
+   *
+   * @throws SearchIndexException
+   *          Thrown if unable to add or update the series.
+   */
+  public List<Optional<Series>> addOrUpdateSeriesBundle(List<String> ids,
+          List<Function<Optional<Series>, Optional<Series>>> updateFunctionBundle,
+          String orgId, User user) throws SearchIndexException {
+    List<Lock> lockBundle = new ArrayList<>();
+    for (String id: ids) {
+      lockBundle.add(this.locks.get(id));
+      lockBundle.get(lockBundle.size() - 1).lock();
+      logger.debug("Locked series '{}'", id);
+    }
+    try {
+      List<Optional<Series>> updatedSeriesOptBundle = new ArrayList<>();
+      List<Series> updatedSeriesBundle = new ArrayList<>();
+      boolean allPresent = true;
+      for (int i = 0; i < ids.size(); i++) {
+        Optional<Series> seriesOpt = getSeries(ids.get(i), orgId, user,
+                maxRetryAttemptsUpdate, retryWaitingPeriodUpdate);
+        updatedSeriesOptBundle.add(updateFunctionBundle.get(i).apply(seriesOpt));
+        if (!updatedSeriesOptBundle.get(updatedSeriesOptBundle.size() - 1).isPresent()) {
```

Comment:
Isn't this basically guaranteed to be `false` all the time, since we just run a function on this which will update or create that series? And assuming this could be `true`, that would mean, we would just skip updating the whole block of series?

Reply:
(same as above)